### PR TITLE
*: Fix issue #2175

### PR DESCRIPTION
**Pull request by zimulala**
Fix issue #2175

### CODE REVIEW

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -127,6 +127,10 @@ type SessionVars struct {
 
 	// StmtCtx holds variables for current executing statement.
 	StmtCtx *StatementContext
+
+	// CurrValues is used to record current ValuesExpr's values.
+	// See http://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_values
+	CurrValues interface{}
```
Comment (coocood) [inline sessionctx/variable/session.go]: Rename to `CurrInsertRow` is better.
`CurrInsertRow` just contains the value of `ValueExpr`,  not the values of the `ValueExpr`.

File: evaluator/evaluator.go [hunk 2]
```diff
@@ -582,6 +582,16 @@ func (e *Evaluator) unaryOperation(u *ast.UnaryOperationExpr) bool {
...
+	values := e.ctx.GetSessionVars().CurrValues
+	if values != nil {
+		row := values.([]types.Datum)
+		off := v.Column.Refer.Column.Offset
+		if len(row) > off {
+			v.SetDatum(row[off])
+			return true
+		}
+	}
+
```
Comment (coocood) [inline evaluator/evaluator.go]: We should set error on `Evaluator` if currValues is nil.

File: executor/executor_write.go [hunk 3]
```diff
@@ -617,6 +618,10 @@ func (e *InsertExec) Next() (*Row, error) {
 			}
 			return nil, errors.Trace(err)
 		}
+
+		// For evaluating ValuesExpr
+		// See http://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_values
+		e.ctx.GetSessionVars().CurrValues = row
```
Comment (coocood) [inline executor/executor_write.go]: Move this line into `onDeuplicate` is better, to keep related logic together.

**Comment by zimulala** [comment]
PTAL @coocood

File: executor/executor_write.go [hunk 4]
```diff
@@ -908,20 +910,22 @@ func (e *InsertValues) initDefaultValues(row []types.Datum, marked map[int]struc
...
-	// Evaluate the updated value.
-	// TODO: report rows affected and last insert id.
 	data, err := e.Table.Row(e.ctx, h)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	// For evaluate ValuesExpr
-	// http://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_values
+
+	// for evaluating value
```
Comment (coocood) [inline executor/executor_write.go]: for evaluating `ColumnNameExpr`.

**Comment by coocood** [comment]
LGTM

File: executor/executor_write.go [hunk 5]
```diff
@@ -597,6 +597,7 @@ func (e *InsertExec) Next() (*Row, error) {
 		return nil, errors.Trace(err)
 	}
 
+	defer func() { e.ctx.GetSessionVars().CurrInsertValues = nil }()
```
Comment (hanfei1991) [inline executor/executor_write.go]: Can we leave this clean work to Close function ?

File: evaluator/evaluator.go [hunk 6]
```diff
@@ -582,8 +582,21 @@ func (e *Evaluator) unaryOperation(u *ast.UnaryOperationExpr) bool {
...
-	v.SetDatum(*v.Column.GetDatum())
-	return true
+	values := e.ctx.GetSessionVars().CurrInsertValues
+	if values == nil {
+		e.err = errors.New("Session current insert values is nil")
+		return false
+	}
+
+	row := values.([]types.Datum)
+	off := v.Column.Refer.Column.Offset
```
Comment (hanfei1991) [inline evaluator/evaluator.go]: off -> offset

File: evaluator/evaluator.go [hunk 7]
```diff
@@ -582,8 +582,21 @@ func (e *Evaluator) unaryOperation(u *ast.UnaryOperationExpr) bool {
...
+	if len(row) > off {
+		v.SetDatum(row[off])
+		return true
+	}
+
+	e.err = errors.Errorf("Session current insert values len %d and column's offset %v don't match", len(row), off)
```
Comment (hanfei1991) [inline evaluator/evaluator.go]: When this error would be reported?

File: executor/executor_write.go (hunk 5 above)
Comment (hanfei1991) [inline executor/executor_write.go]: Why we need this ?

File: evaluator/evaluator.go (hunk 7 above)
Comment (zimulala) [inline evaluator/evaluator.go]: Yep.

File: executor/executor_write.go (hunk 5 above)
Comment (zimulala) [inline executor/executor_write.go]: Clean `CurrInsertValues`  when the end of inserting.

Comment (hanfei1991) [inline executor/executor_write.go]: Why should we clean it for every Next?

Comment (zimulala) [inline executor/executor_write.go]: `Insert` only calls `Next` once, but put it to `Close` is better.

Comment (zimulala) [inline executor/executor_write.go]: PTAL @hanfei1991

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**